The polymer is formed by smaller units which, when triggered, react with each other such that two adjacent units of the same type and opposite polarity are destroyed. Units' types are represented by letters; units' polarity is represented by capitalization. For instance, r and R are units with the same type but opposite polarity, whereas r and s are entirely different types and do not react.

For example:

- In aA, a and A react, leaving nothing behind.
- In abBA, bB destroys itself, leaving aA. As above, this then destroys itself, leaving nothing.
- In abAB, no two adjacent units are of the same type, and so nothing happens.
- In aabAAB, even though aa and AA are of the same type, their polarities match, and so nothing happens.

## Part 1

How many units remain after fully reacting the polymer you scanned?

In [1]:
import numpy as np
import pandas as pd
import re

In [6]:
input_str = open("input.txt").read().splitlines()[0]

In [38]:
nstr = list(input_str)
while True:
    comp_str = ''.join(nstr)
    for i in range(0,len(nstr) - 1):
        char1 = nstr[i]
        char2 = nstr[i + 1]
        if (char1.islower() and char2.isupper()) or (char2.islower() and char1.isupper()):
            if char1.capitalize() == char2.capitalize():
                nstr[i], nstr[i+1] = '', ''
    res = ''.join(nstr)
    if res == comp_str:
        #print(res)
        print(len(res))
        break
    else:
        nstr = list(res)


10878


### Part 1 answer is 10878

## Part 2

--- Part Two ---
Time to improve the polymer.

One of the unit types is causing problems; it's preventing the polymer from collapsing as much as it should. Your goal is to figure out which unit type is causing the most problems, remove all instances of it (regardless of polarity), fully react the remaining polymer, and measure its length.

In [36]:
unique_chars_lower = list(set(input_str.lower()))

In [41]:
result_lengths = []
original_list = list(input_str)
for i in range(0, len(unique_chars_lower) - 1):
    char_to_remove = unique_chars_lower[i]
    nlist = [c.replace(char_to_remove, '') for c in original_list]
    nlist = [c.replace(char_to_remove.capitalize(), '') for c in nlist]
    nstr = list(''.join(nlist))
    flag = True
    while flag:
        comp_str = ''.join(nstr)
        for i in range(0,len(nstr) - 1):
            char1 = nstr[i]
            char2 = nstr[i + 1]
            if (char1.islower() and char2.isupper()) or (char2.islower() and char1.isupper()):
                if char1.capitalize() == char2.capitalize():
                    nstr[i], nstr[i+1] = '', ''
        res = ''.join(nstr)
        if res == comp_str:
            result_lengths.append(len(res))
            flag = False
        else:
            nstr = list(res)
min(result_lengths)

6874

### Answer to Part 2 is 6874